full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,166 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+/* eslint-disable @typescript-eslint/no-empty-function -- Remove me */
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+import { isDOMComponent } from 'react-dom/test-utils';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function getTodos() {
+      try {
+        const response = await fetch(`/api/todos`);
+        if (!response.ok) {
+          throw new Error(`there was an error of ${response.status}`);
+        }
+        const todos = await response.json();
+        setTodos(todos);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    getTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const response = await fetch('/api/todos', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newTodo),
+      });
+      if (!response.ok) {
+        throw new Error(`there was an error of ${response.status}`);
+      }
+      const addedTodo = await response.json();
+      setTodos((prevTodo) => [...prevTodo, addedTodo]);
+      return;
+    } catch (err) {
+      setError(err);
+    } finally {
+      setIsLoading(false);
+    }
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todo: Todo) {
+    try {
+      todo.isCompleted = !todo.isCompleted;
+
+      const response = await fetch(`/api/todos/${todo.todoId}`, {
+        method: 'PUT',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(todo),
+      });
+      if (!response.ok) {
+        throw new Error(`there was an error of ${response.status}`);
+      }
+      const updatedTodo = await response.json();
+      console.log(updatedTodo);
+      setTodos((prevTodos) => [
```

Comment:
Here, also, you could simplify this to
`setTodos(todos.map((todo) => todo.todoId === updatedTodo.todoId ? updatedTodo : todo));`